File: algorithm/dfs/dfs.cpp
#include<iostream>
#include<list>

using namespace std;

// maximum graph nodes
#define MAX_NODES 1000

// graph adjacency list
list<int> graph[MAX_NODES];

// nodes color
short color[MAX_NODES];

// graph nodes count
int nodes;

/**
 *  init
 */
void init(){
    for (int i=0; i<MAX_NODES; i++) {
        color[i] = 0;
    }
}

/**
 *  Visit each node and every descendants
 */
void dfs_visit(int i){
    color[i] = 1;
    int value;

    // visited

    list<int>::iterator it = graph[i].begin();
    for (; it != graph[i].end(); it++) {
        value = (int) *it;
        if (color[value] == 0) {
            dfs_visit(value);
        }
    }
    
    // processed
}

/**
 *  Call dfs_visit for every unvisited node
 */
void dfs(){
    for (int i=0; i<nodes; i++) {
        if (color[i] == 0) {
            dfs_visit(i);
        }
    }
}

File: algorithm/dfs/DFS.java
import java.util.ArrayList;

public class DFS {

    // adjacency list
    ArrayList<Integer> graph[];

    // nodes color
    byte colors[];

    /**
     * initialize dfs
     * @param n number of graph nodes
     */
    public void init(int n){
        graph = new ArrayList[n];
        colors = new byte[n];
        for (int i = 0; i < n; i++){
            graph[i] = new ArrayList<Integer>();
        }
    }

    /**
     * Start DFS search
     */
    public void dfs(){
        for (int i = 0; i < graph.length; i++) {
            if (colors[i] == 0) {
                dfsVisit(i);
            }
        }
    }

    /**
     * Run DFS for a node
     * @param index start index of DFS
     */
    private void dfsVisit(int index) {
        colors[index] = 1;
        for (int i = 0; i < graph[index].size(); i++) {
            if (colors[graph[index].get(i)] == 0) {
                dfsVisit(i);
            }
        }
    }
}
